### ISSUE TITLE
Restart multiple instantiations of a MultiApp from checkpoint

### ISSUE THREAD

**Issue by vincentlaboure** [idaholab/moose#15287]
## Reason

A typical multiphysics setup is to first obtain a multiapp steady-state solution and to restart from that solution as the initial condition for a transient calculation. Checkpoint files are often used to restart.

If the master app only instantiates a given sub-app once, then the sub-apps can be restarted as follows (see [here](https://github.com/idaholab/moose/tree/next/test/tests/restart/restart_transient_from_steady) for more details):

```
[Mesh]
  file = master_out_subapp0_checkpoint_cp/LATEST
[]


[Problem]
  restart_file_base = master_out_subapp0_checkpoint_cp/LATEST
  skip_additional_restart_data = true
  force_restart = true
[]
``` 

However, if that sub-app is instantiated in multiple locations, then multiple `master_out_subapp*_checkpoint_cp` checkpoint folders will be created and it is not currently possible to have a single transient sub-app input file that will pick up the correct folder based on location.

## Design

Ideally, it would be great to have something like this in the transient sub-app input file:

```
[Mesh]
  file = master_out_subapp*_checkpoint_cp/LATEST
[]


[Problem]
  restart_file_base = master_out_subapp*_checkpoint_cp/LATEST
  skip_additional_restart_data = true
  force_restart = true
[]
``` 

and that the sub app automatically understands to get the folder with '*' corresponding to the instantiation index of the sub-app. Or something to that effect.

## Impact

This feature seems essential to many multiphysics problems with transients restarting from steady-states.

**Comment by friedmud** [comment]
This is not really how this is supposed to work.  The checkpoint for all of the sub-apps is "rolled up" into the checkpoint for the master app.  When you restart a MultiApp system the checkpoints for the sub-apps are "unpacked" from the master app's checkpoint and pushed down.

There should be no need to mess with the checkpoint for the sub-apps.  In fact - you shouldn't even be writing checkpoint from the sub-apps.  Only write it from the master app.

**Comment by friedmud** [comment]
(Also: if restarting like this working - please let us know!)

**Comment by vincentlaboure** [comment]
I did see this comment [in the documentation](https://mooseframework.inl.gov/application_usage/restart_recover.html): 

```
MultiApp Restart
When running a multiapp simulation you do not need to enable checkpoint output in each sub app input file. The master app stores the restart data for all sub apps in its file.
```

which seems to be what you are explaining here.

However, I was confused because I then looked at the tests in moose and saw that checkpoint files were used for both the master and the sub-app (e.g. [here](https://github.com/idaholab/moose/tree/next/test/tests/restart/restart_transient_from_steady)).

@friedmud Could you please point me to an example where the master checkpoint is used to restart the sub-apps?

**Comment by fdkong** [comment]
> This is not really how this is supposed to work. The checkpoint for all of the sub-apps is "rolled up" into the checkpoint for the master app. When you restart a MultiApp system the checkpoints for the sub-apps are "unpacked" from the master app's checkpoint and pushed down.
> 
> There should be no need to mess with the checkpoint for the sub-apps. In fact - you shouldn't even be writing checkpoint from the sub-apps. Only write it from the master app.

It is a beautiful solution! I like it!

**Comment by YaqiWang** [comment]
@vincentlaboure can you also include the syntax for restarting master? This sounds to me that it is the master who orders for restart for multiapps but not let the multiapps restart on their own.

**Comment by vincentlaboure** [comment]
@YaqiWang The master would restart with something like:

```
[Mesh]
  file = master_out_checkpoint_cp/LATEST
[]

[Problem]
  restart_file_base = master_out_checkpoint_cp/LATEST
  skip_additional_restart_data = true # omit if restarting from Transient
  force_restart = true
[]
```

I just don't know how the sub-apps would know how to restart. Is it done automatically?

**Comment by vincentlaboure** [comment]
As a follow-up, I am not able to restart master+sub-app properly if I'm not using checkpoints at the sub-app level. But it could just be that I don't know how.

**Comment by tophmatthews** [comment]
Can you get a working example to put in as a test with only using moose @vincentlaboure ? It may be helpful to see a non-working example

**Comment by vincentlaboure** [comment]
Now here is something very odd to me. If I keep the following:

```
[Problem]
  restart_file_base = master_out_subapp0_checkpoint_cp/LATEST
  skip_additional_restart_data = true
  force_restart = true
[]
```

**without** using the `cli_args` like @permcody suggested, all the sub apps do restart from the desired values... 😮 Same if I use `restart_file_base = master_out_subapp5_checkpoint_cp/LATEST`. Could it be that the restart data for all the sub-apps is accumulated into each sub-app checkpoint folder?

It's great news to me because it means that I am able to establish the initial condition I want (except that all but one of the sub-app checkpoint folders seem redundant), but is it expected behavior, @friedmud?

**Comment by tophmatthews** [comment]
Note, I've had to turn off force_restart to subapps in the past to do recovers, hopefully that doesn't screw this up here too?
```
moose_test-opt -i initial_master.i
moose_test-opt -i transient_master.i Executioner/num_steps=2
moose_test-opt -i transient_master.i temp_subapp0:Problem/force_restart=false --recover
```

**Comment by YaqiWang** [comment]
What @friedmud said could be this test? 
```
moose/test/tests/multiapps/restart_subapp_ic]> ../../../moose_test-opt -i master2.i
```
If you comment out the initial condition on `u` in `sub2.i`, we seems restarted multapp without touching its input. I guess MOOSE creates a backup of the subapp in memory and outputs into master checkpoint file. Thus when restarting, the multiapp is restored from that backup loaded from the master. @vincentlaboure Is this what you want? That test is not restarting a transient from a steady checkpoint file though.

**Comment by YaqiWang** [comment]
I second @tophmattews suggestion: design a moose test (restarting a transient from a steady checkpoint file all have multiple instances of multiapps).

**Comment by fdkong** [comment]
> However, I was confused because I then looked at the tests in moose and saw that checkpoint files were used for both the master and the sub-app (e.g. [here](https://github.com/idaholab/moose/tree/next/test/tests/restart/restart_transient_from_steady)).

I got these flies from you @vincentlaboure . It looks like we should redo these tests according to @friedmud 's comments. I think it is the right solution to let master handle everything.

**Comment by vincentlaboure** [comment]
As a follow-up, I was able to get @friedmud's suggestion to work (not sure what I did different when I previously reported that I wasn't able to), restarting both Transient from Transient and Steady from Transient with a single checkpoint.

Before closing the issue, I will create a few MOOSE tests making sure that this is tested with multiple instantiations of a sub app.

Thanks @friedmud, @YaqiWang, @tophmatthews and @fdkong for the help!
